diff --git a/modules/comment/comment.module b/modules/comment/comment.module
index a83069f..98f09e0 100644
--- a/modules/comment/comment.module
+++ b/modules/comment/comment.module
@@ -2193,17 +2193,20 @@ function comment_submit($comment) {
     // 2) Strip out all HTML tags
     // 3) Convert entities back to plain-text.
     $field = field_info_field('comment_body');
-    $langcode = field_is_translatable('comment', $field) ? entity_language('comment', $comment) : LANGUAGE_NONE;
-    $comment_body = $comment->comment_body[$langcode][0];
-    if (isset($comment_body['format'])) {
-      $comment_text = check_markup($comment_body['value'], $comment_body['format']);
-    }
-    else {
-      $comment_text = check_plain($comment_body['value']);
+    // Do not try to extract from comment_body if no such field exists.
+    if ($field && isset($comment->comment_body)) {
+      $langcode = field_is_translatable('comment', $field) ? entity_language('comment', $comment) : LANGUAGE_NONE;
+      $comment_body = $comment->comment_body[$langcode][0];
+      if (isset($comment_body['format'])) {
+        $comment_text = check_markup($comment_body['value'], $comment_body['format']);
+      }
+      else {
+        $comment_text = check_plain($comment_body['value']);
+      }
+      $comment->subject = truncate_utf8(trim(decode_entities(strip_tags($comment_text))), 29, TRUE);
     }
-    $comment->subject = truncate_utf8(trim(decode_entities(strip_tags($comment_text))), 29, TRUE);
-    // Edge cases where the comment body is populated only by HTML tags will
-    // require a default subject.
+    // Edge cases where the comment body is populated only by HTML tags or does
+    // not exist will require a default subject.
     if ($comment->subject == '') {
       $comment->subject = t('(No subject)');
     }
diff --git a/modules/comment/comment.test b/modules/comment/comment.test
index 30bff71..44e95ac 100644
--- a/modules/comment/comment.test
+++ b/modules/comment/comment.test
@@ -2102,6 +2102,25 @@ class CommentFieldsTest extends CommentHelperCase {
     $edit = array('comment_body[und][0][value]' => $this->randomName(8));
     $this->drupalPost('node/' . $this->node->nid, $edit, t('Save'));
   }
+
+  /**
+   * Test that certain comment functionalities work without the comment_body
+   * field, and without generating warnings.
+   */
+  function testDeletedCommentBody() {
+    // Delete the comment_body field.
+    $this->drupalLogin($this->admin_user);
+    $this->drupalPost('admin/structure/types/manage/article/comment/fields/comment_body/delete', array(), t('Delete'));
+
+    // Preview an empty comment.
+    $this->drupalPost('node/' . $this->node->nid, array(), t('Preview'));
+
+    // Post an empty comment.
+    $this->drupalPost('node/' . $this->node->nid, array(), t('Save'));
+
+    // View the comment administration overview page.
+    $this->drupalGet('admin/content/comment');
+  }
 }
 
 /**
