diff --git a/field_collection_table.module b/field_collection_table.module
index 26e64e8..a031a58
--- a/field_collection_table.module
+++ b/field_collection_table.module
@@ -56,7 +56,7 @@ function field_collection_table_field_formatter_view($entity_type, $entity, $fie
         $view = $field_collection->view();
         $content = $view['field_collection_item'][$field_collection->identifier()];
         $column = array();
-        foreach ($field_names as $field_name) {
+        foreach ($field_names as $weight => $field_name) {
           if (isset($content[$field_name])) {
             $content[$field_name]['#label_display'] = 'hidden';
           }
@@ -70,6 +70,9 @@ function field_collection_table_field_formatter_view($entity_type, $entity, $fie
             'data' => $content[$field_name],
             'class' => $field_name,
           );
+          if (isset($content[$field_name]['#title'])) {
+            $header[$weight]['data'] = $content[$field_name]['#title'];  
+          }
         }
 
         $links = array(
