diff --git a/file_entity.module b/file_entity.module
index 9f5a974..2f685a1 100644
--- a/file_entity.module
+++ b/file_entity.module
@@ -1890,7 +1890,7 @@ function _file_entity_query_file_entity_access_alter($query, $type) {
           $base_alias = $falias;
           $field = 'entity_id';
         }
-        $file_managed_alias = $query->join('file_managed', 'fm', "fm.fid = $falias.$field");
+        $file_managed_alias = $query->leftJoin('file_managed', 'fm', "fm.fid = $falias.$field");
       }
       else {
         $file_managed_alias = $falias;
