From c6c7c228d0fec3cbc58b5f6cb6aaf5824ea6e99f Mon Sep 17 00:00:00 2001
From: Thomas Fini Hansen <xen@xen.dk>
Date: Tue, 20 Mar 2012 07:53:53 +0100
Subject: [PATCH] Clear cache when refreshing projects.

---
 l10n_update.project.inc |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/l10n_update.project.inc b/l10n_update.project.inc
index 71a3b3d..505d9af 100644
--- a/l10n_update.project.inc
+++ b/l10n_update.project.inc
@@ -140,6 +140,7 @@ function l10n_update_project_refresh($modules) {
   $projects = array();
 
   // Get all current projects, including the recently installed.
+  l10n_update_get_projects(TRUE);
   $current_projects = l10n_update_build_projects(TRUE);
   // Collect project data of newly installed projects.
   foreach ($modules as $name) {
-- 
1.7.5.4

